python/dgl/graphbolt/minibatch.py

```
@@ -433,25 +433,52 @@ def node_pairs_with_labels(self):
             return x.to(device) if hasattr(x, "to") else x
 
-        for attr in dir(self):
+        def apply_to(x, device):
+            return recursive_apply(x, lambda x: _to(x, device))
+
+        if self.labels is not None and self.seed_nodes is not None:
+            # Node classification.
+            transfer_attrs = [
+                "labels",
+                "sampled_subgraphs",
+                "node_features",
+                "edge_features",
+            ]
+        elif self.labels is None and self.compacted_node_pairs is not None:
+            # Link prediction.
+            transfer_attrs = [
+                "compacted_node_pairs",
+                "compacted_negative_srcs",
+                "compacted_negative_dsts",
+                "sampled_subgraphs",
+                "node_features",
+                "edge_features",
+            ]
+        elif self.labels is not None and self.compacted_node_pairs is not None:
+            # Edge classification.
+            transfer_attrs = [
+                "labels",
+                "compacted_node_pairs",
+                "sampled_subgraphs",
+                "node_features",
+                "edge_features",
+            ]
+        else:
+            # Unrecognized task. Will copy all the attributes to the device.
+            transfer_attrs = get_attributes(self)
+
+        for attr in transfer_attrs:
             # Only copy member variables.
-            if not callable(getattr(self, attr)) and not attr.startswith("__"):
-                try:
-                    setattr(
-                        self,
-                        attr,
-                        recursive_apply(
-                            getattr(self, attr), lambda x: _to(x, device)
-                        ),
-                    )
-                except AttributeError:
-                    continue
+            try:
+                setattr(self, attr, apply_to(getattr(self, attr), device))
+            except AttributeError:
```

as discussed offline, we'd better figure out a way to avoid such except.

---

add comments here to explain why we swallow exception silently.